File: python3/backtracking/0077.py

from typing import List

class Solution:
    def combine(self, n: int, k: int) -> List[List[int]]:
        def dfs(cur):
            if len(path) + (n - cur + 1) < k:
                return
            
            if len(path) == k:
                rst.append(path.copy())
                return

            for i in range(cur, n+1):
                path.append(i)
                dfs(i+1)
                path.pop()
            
            return

        path = []
        rst = []
        dfs(1)
        return rst


def testCombinationSum():
    test_case = [
        [4, 2],
    ]

    print('----------------- Test ------------------')
    solu = Solution()
    for item in test_case:
        # import ipdb; ipdb.set_trace()
        rst = solu.combine(item[0], item[1])
        print('result = {}\n'.format(rst))


if __name__ == '__main__':
    testCombinationSum()

File: cpp/0077.Combinations/0077.cpp
#include <iostream>
#include <vector>
using namespace std;

class Solution {
public:
    vector<vector<int>> combine(int n, int k) {
        vector<int> path;
        vector<vector<int>> rst;

        dfs(1, n, k, path, rst);

        return rst;
    }

    void dfs(int cur, int n, int k, vector<int>& path, vector<vector<int>>& rst) {
        if (path.size() + (n-cur+1) < k) {
            return;
        }

        if (path.size() == k) {
            rst.push_back(path);
            return;
        }

        path.push_back(cur);
        dfs(cur+1, n, k, path, rst);
        path.pop_back();

        dfs(cur+1, n, k, path, rst);

    }

};


int main() {

    cout << "============= Test ==============" << endl;
    
    Solution solu;
    auto rst = solu.combine(4, 2);

    
    return 0;
}

